File: WorkcraftCore/src/main/java/org/workcraft/gui/PropertyEditorWindow.java
/*
*
* Copyright 2008,2009 Newcastle University
*
* This file is part of Workcraft.
* 
* Workcraft is free software: you can redistribute it and/or modify
* it under the terms of the GNU General Public License as published by
* the Free Software Foundation, either version 3 of the License, or
* (at your option) any later version.
* 
* Workcraft is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
* 
* You should have received a copy of the GNU General Public License
* along with Workcraft.  If not, see <http://www.gnu.org/licenses/>.
*
*/

package org.workcraft.gui;

import java.awt.BorderLayout;

import javax.swing.JPanel;
import javax.swing.JScrollPane;

import org.workcraft.Framework;
import org.workcraft.dependencymanager.advanced.core.EvaluationContext;
import org.workcraft.dependencymanager.advanced.core.Expression;
import org.workcraft.dependencymanager.advanced.core.Expressions;
import org.workcraft.dependencymanager.advanced.user.AutoRefreshExpression;
import org.workcraft.dependencymanager.advanced.user.Variable;
import org.workcraft.gui.propertyeditor.EditableProperty;
import org.workcraft.gui.propertyeditor.PropertyEditorTable;

import pcollections.PVector;
import pcollections.TreePVector;

@SuppressWarnings("serial")
public class PropertyEditorWindow extends JPanel {

	private PropertyEditorTable propertyTable;
	private JScrollPane scrollProperties;
	@SuppressWarnings("unused")//The field is used to keep reference to an otherwise garbage-collected refresher
	private final AutoRefreshExpression refresher;
	
	public PropertyEditorWindow (Framework framework) {
		propertyTable = new PropertyEditorTable(framework);

		scrollProperties = new JScrollPane();
		scrollProperties.setViewportView(propertyTable);

		setLayout(new BorderLayout(0,0));
		add(new DisabledPanel(), BorderLayout.CENTER);
		validate();
		
		refresher = new AutoRefreshExpression() {
			@Override
			protected void onEvaluate(EvaluationContext context) {
				PVector<EditableProperty> obj = context.resolve(prop);
				if(obj == null)
					clearObject();
				else
					setObject(obj);
			}
		};
	}
	
	public final Variable<Expression<? extends PVector<EditableProperty>>> propertyObject = new Variable<Expression<? extends PVector<EditableProperty>>>(Expressions.constant(TreePVector.<EditableProperty>empty()));
	final Expression<PVector<EditableProperty>> prop = Expressions.join(propertyObject);
	
	public void setObject (PVector<EditableProperty> o) {
		removeAll();
		propertyTable.setObject(o);
		add(scrollProperties, BorderLayout.CENTER);
		validate();
		repaint();
	}

	public void clearObject () {
		removeAll();
		propertyTable.clearObject();
		add(new DisabledPanel(), BorderLayout.CENTER);
		validate();
		repaint();
	}
}

File: Gui/src/main/scala/org/workcraft/gui/propertyeditor/PropertyEditorWindow.scala
package org.workcraft.gui.propertyeditor

import java.awt.BorderLayout
import javax.swing.JPanel
import javax.swing.JScrollPane
import pcollections.PVector
import pcollections.TreePVector
import org.workcraft.dependencymanager.advanced.user.AutoRefreshExpression
import org.workcraft.dependencymanager.advanced.core.EvaluationContext
import org.workcraft.dependencymanager.advanced.user.Variable
import org.workcraft.scala.Expressions._
import org.workcraft.scala.effects.IO._
import scalaz._
import Scalaz._

class PropertyEditorWindow extends JPanel {
  private val propertyTable: PropertyEditorTable = new PropertyEditorTable
  private val scrollProperties: JScrollPane = new JScrollPane
  scrollProperties.setViewportView(propertyTable)
  setLayout(new BorderLayout(0, 0))
  add(new DisabledPanel(), BorderLayout.CENTER)
  validate()
  val propertyObject: Variable[Expression[Option[List[Expression[EditableProperty]]]]] = Variable.create(constant(None))

  private val prop: Expression[Option[List[EditableProperty]]] = (propertyObject: Expression[Expression[Option[List[Expression[EditableProperty]]]]]).join >>= (_.traverse(_.sequence))

  //@SuppressWarnings(Array("unused"))

  private val refresh = swingAutoRefresh(prop, (p: Option[List[EditableProperty]]) => ioPure.pure {
    p match {
      case None => clearObject
      case Some(props) => setObject(props)
    }
  })

  private val refresher: AutoRefreshExpression = new AutoRefreshExpression {
    override def onEvaluate(context: EvaluationContext) = {
      val obj = context.resolve(prop)
    }
  }

  def setObject(o: List[EditableProperty]): Unit = {
    removeAll()
    propertyTable.setObject(o)
    add(scrollProperties, BorderLayout.CENTER)
    validate()
    repaint()
  }

  def clearObject(): Unit = {
    removeAll()
    propertyTable.clearObject()
    add(new DisabledPanel, BorderLayout.CENTER)
    validate()
    repaint()
  }
}
